packages/traceability/issuer-irec-api/src/pods/certificate/handler/import-irec-certificate.handler.ts

```diff
@@ -0,0 +1,99 @@
+import { BadRequestException, Inject, InternalServerErrorException } from '@nestjs/common';
+
+import { IREC_SERVICE, IrecService } from '@energyweb/origin-organization-irec-api';
+import { DeviceService } from '@energyweb/origin-device-registry-irec-local-api';
+import { UserService } from '@energyweb/origin-backend';
+
+import { ImportIrecCertificateCommand } from '../command';
+import { IrecCertificateImportFailedEvent } from '../event';
+import {
+    ApproveCertificationRequestCommand,
+    CreateCertificationRequestCommand
+} from '@energyweb/issuer-api';
+import { DeviceRegistryService } from '@energyweb/origin-device-registry-api';
+import { Repository } from 'typeorm';
+import { InjectRepository } from '@nestjs/typeorm';
+import { IrecCertificationRequest } from '../../certification-request';
+
+@CommandHandler(ImportIrecCertificateCommand)
+export class ImportIrecCertificateHandler implements ICommandHandler<ImportIrecCertificateCommand> {
+    constructor(
+        private readonly queryBus: QueryBus,
+        private readonly eventBus: EventBus,
+        private readonly commandBus: CommandBus,
+        @Inject(IREC_SERVICE)
+        private readonly irecService: IrecService,
+        private readonly irecDeviceService: DeviceService,
+        private readonly deviceRegistryService: DeviceRegistryService,
+        private readonly userService: UserService,
+        @InjectRepository(IrecCertificationRequest)
+        private readonly irecCertificationRequestRepository: Repository<IrecCertificationRequest>
+    ) {}
+
+    async execute({ user, certificateToImport }: ImportIrecCertificateCommand): Promise<void> {
+        const irecCertificates = await this.irecService.getCertificates(user);
+        const irecCertificate = irecCertificates.find(
+            (c) => c.asset === certificateToImport.assetId
+        );
+
+        if (!irecCertificate) {
+            throw new BadRequestException(
+                `IREC certificate for asset ${certificateToImport.assetId} not found`
+            );
+        }
+
+        const irecDevices = await this.irecDeviceService.findAll({
+            where: { ownerId: user.ownerId }
+        });
+        const irecDevice = irecDevices.find((d) => d.code === irecCertificate.device.code);
+        if (!irecDevice || irecDevice.ownerId !== user.ownerId) {
+            throw new BadRequestException('Unknown IREC device');
+        }
+
+        const [originDevice] = await this.deviceRegistryService.find({
+            where: { externalRegistryId: irecDevice.id }
+        });
+        if (!originDevice) {
+            throw new BadRequestException('Unknown IREC device');
+        }
+
+        const platformAdmin = await this.userService.getPlatformAdmin();
+        const transaction = await this.irecService.transferCertificate(
+            user,
+            platformAdmin.organization.id,
+            certificateToImport.assetId
+        );
```

Comment:
Is this where if we check if the user owns the asset it's trying to import?

Reply:
ownership check is on line 40